### DISCUSSION TITLE
@ConfigProperty being set despite bean being disabled via @IfBuildProperty

### DISCUSSION THREAD

**Discussion by wiebeck** [quarkusio/quarkus#37264] [Q&A]
My Quarkus project can be built as "server" or "worker" binary depending on the `mode` system property I give on the Gradle build commandline (example: `-Pmode=worker`). The server binary should contain REST endpoints and a custom `JWTCallerPrincipalFactory` for validating JWTs. This `JWTCallerPrincipalFactory` shall not be enabled for the worker binary because the worker does not provide REST endpoints. This is how the `JWTCallerPrincipalFactory` bean looks like (just an excerpt):

```java
@ApplicationScoped
@Alternative
@Priority(1)
@IfBuildProperty(name = "mode", stringValue = "server")
public class MyJWTCallerPrincipalFactory extends JWTCallerPrincipalFactory {

    @ConfigProperty(name = "server.property")
    String someProperty;
```

The `server.property` only has a value for the server build but none for the worker build. Since we updated the project to Quarkus 3 today we noticed that the **worker binary** doesn't start up properly any more (the server is running fine). The error is:

> Failed to load config value of type class java.lang.String for: server.property

So Quarkus 3 actually tries to set the properties to some values although there aren't any. Quarkus 2 did not complain about the missing value, either because it simply set `null` in this case or it actually didn't even try to set properties because the bean has been disabled by the `@IfBuildProperty` annotation.

Is this a known issue? I couldn't find anything in the Quarkus 3 migration guide about this. We're currently trying to set `defaultValues` for these properties, which should resolve the issue. But I still wonder why Quarkus tries to set properties on a bean that has been disabled during build time. Sounds like a bug to me.

**Comment by wiebeck** [comment]
I found out that `@IfBuildPropery` doesn't work as expected when I try to set the property as Gradle commandline parameter either via `-Dserver.property=foo` or `-Pserver.property=foo`.

What _DOES_ work on the other hand is setting the build property as environment variable:

```
SERVER_PROPERTY=foo ./gradlew build
```

This definitely changed in Quarkus 3. I will try to find some time and create a reproducible repository.

**Reply by mkouba** [reply]
> So Quarkus 3 actually tries to set the properties to some values although there aren't any. Quarkus 2 did not complain about the missing value, either because it simply set null in this case or it actually didn't even try to set properties because the bean has been disabled by the @IfBuildProperty annotation.

What Quarkus version do you use?

> This definitely changed in Quarkus 3. I will try to find some time and create a reproducible repository.

A reproducer would be really useful.
